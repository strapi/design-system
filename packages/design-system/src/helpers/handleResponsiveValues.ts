import { DefaultTheme } from 'styled-components';

import { extractStyleFromTheme } from './theme';

import type { TransientBoxProps } from '../primitives/Box/Box';
import type { TransientFlexProps } from '../primitives/Flex/Flex';
import type { TransientTypographyProps } from '../primitives/Typography/Typography';
import type { DefaultThemeOrCSSProp } from '../types';

type Breakpoint = 'initial' | 'small' | 'medium' | 'large';

/**
 * A property is either a responsive object, or a single
 * value that is applied as the initial value.
 */
type ResponsiveProperty<T> =
  | {
      [key in Breakpoint]?: T;
    }
  | T;

/**
 * Currently, only margin or padding accept an array of values.
 */
type ResponsiveThemeProperty<T extends keyof DefaultTheme, K extends keyof React.CSSProperties> = K extends
  | 'padding'
  | 'margin'
  ? ResponsiveProperty<DefaultThemeOrCSSProp<T, K> | Array<DefaultThemeOrCSSProp<T, K>>>
  : ResponsiveProperty<DefaultThemeOrCSSProp<T, K>>;

/**
 * This should ONLY ever be CSS property names, never shorthands or aliases.
 */
type ResponsiveProps = Omit<TransientBoxProps, 'basis' | 'grow' | 'shrink' | 'shadow'> &
  Omit<TransientFlexProps, 'direction' | 'wrap'> &
  Omit<TransientTypographyProps, 'ellipsis' | 'variant'> & {
    boxShadow?: TransientBoxProps['shadow'];
    flexBasis?: TransientBoxProps['basis'];
    flexDirection?: TransientFlexProps['direction'];
    flexGrow?: TransientBoxProps['grow'];
    flexShrink?: TransientBoxProps['shrink'];
    flexWrap?: TransientFlexProps['wrap'];
  };

const mappedCSSProps: Partial<Record<keyof ResponsiveProps, string | string[]>> = {
  padding: ['padding-block-start', 'padding-inline-end', 'padding-block-end', 'padding-inline-start'],
  paddingTop: 'padding-block-start',
  paddingRight: 'padding-inline-end',
  paddingBottom: 'padding-block-end',
  paddingLeft: 'padding-inline-start',
  margin: ['margin-block-start', 'margin-inline-end', 'margin-block-end', 'margin-inline-start'],
  marginLeft: 'margin-inline-start',
  marginRight: 'margin-inline-end',
  marginTop: 'margin-block-start',
  marginBottom: 'margin-block-end',
  borderRadius: 'border-radius',
  borderStyle: 'border-style',
  borderWidth: 'border-width',
  borderColor: 'border-color',
  fontSize: 'font-size',
  fontWeight: 'font-weight',
  lineHeight: 'line-height',
  zIndex: 'z-index',
  boxShadow: 'box-shadow',
  pointerEvents: 'pointer-events',
  textAlign: 'text-align',
  textTransform: 'text-transform',
  textDecoration: 'text-decoration',
  flexGrow: 'flex-grow',
  flexShrink: 'flex-shrink',
  flexBasis: 'flex-basis',
  minWidth: 'min-width',
  maxWidth: 'max-width',
  minHeight: 'min-height',
  maxHeight: 'max-height',
  flexDirection: 'flex-direction',
  flexWrap: 'flex-wrap',
  justifyContent: 'justify-content',
  alignItems: 'align-items',
};

/**
 * Fills the shorthand CSS properties with their corresponding logical properties
 * @param value An array of CSS values for shorthand properties
 * @returns An array of CSS values filled with logical properties
 */

const fillCssValues = (value: DefaultThemeOrCSSProp<'spaces', 'margin' | 'padding'>[]) => {
  const [top, right, bottom, left] = value;
  const rightValue = right ?? top;
  const bottomValue = bottom ?? top;
  const leftValue = left ?? rightValue;

  return [top, rightValue, bottomValue, leftValue] as const;
};

function getThemeSection(key: keyof ResponsiveProps, theme: DefaultTheme) {
  switch (key) {
    case 'gap':
    case 'padding':
    case 'margin':
    case 'paddingTop':
    case 'paddingLeft':
    case 'paddingRight':
    case 'paddingBottom':
    case 'marginTop':
    case 'marginLeft':
    case 'marginRight':
    case 'marginBottom':
    case 'left':
    case 'right':
    case 'top':
    case 'bottom':
    case 'width':
    case 'maxWidth':
    case 'minWidth':
    case 'height':
    case 'maxHeight':
    case 'minHeight':
    case 'borderRadius':
    case 'borderWidth':
      return theme.spaces;
    case 'color':
    case 'background':
    case 'borderColor':
      return theme.colors;
    case 'fontSize':
      return theme.fontSizes;
    case 'fontWeight':
      return theme.fontWeights;
    case 'lineHeight':
      return theme.lineHeights;
    case 'zIndex':
      return theme.zIndices;
    case 'boxShadow':
      return theme.shadows;
    default:
      return null;
  }
}

const handleResponsiveValues = (values: ResponsiveProps, theme: DefaultTheme) => {
  const stylesByBreakpoint = Object.entries(values).reduce(
    (acc, curr) => {
      const [key, value] = curr as [key: keyof ResponsiveProps, value: ResponsiveProps[keyof ResponsiveProps]];
      const themeSection = getThemeSection(key, theme);

      const cssProperty = Object.prototype.hasOwnProperty.call(mappedCSSProps, key) ? mappedCSSProps[key] : key;

      if (cssProperty && (value || value === 0)) {
        // If the value is an responsive object e.g padding : { initial: 1, medium: 2, large: [3, 4] }
        if (typeof value === 'object' && !Array.isArray(value)) {
          Object.entries(value).forEach(([breakpointName, breakpointValue]) => {
            acc[breakpointName] = {
              ...acc[breakpointName],
              ...convertCssPropertiesToCssValues(cssProperty, breakpointValue, themeSection),
            };
          });
        } else {
          /**
           * If we don't pass a responsive object, we just set the value to respective prop.
           */
          acc.initial = {
            ...acc.initial,
            ...convertCssPropertiesToCssValues(cssProperty, value, themeSection),
          };
        }
      }

      return acc;
    },
    {
      initial: {},
      small: {},
      medium: {},
      large: {},
    },
  );

  /**
   * Stringify our styles.
   */
  return Object.entries(stylesByBreakpoint)
    .reduce<string[]>((acc, [key, value]) => {
      if (value && Object.keys(value).length > 0) {
        const breakpointStyles = Object.entries(value)
          .reduce<string[]>((arr, [property, value]) => {
            arr.push(`${property}: ${value};`);

            return arr;
          }, [])
          .join('\n');

        if (key === 'initial') {
          acc.push(breakpointStyles);
        } else {
          acc.push(`${theme.breakpoints[key]}{ ${breakpointStyles} }`);
        }
      }

      return acc;
    }, [])
    .join('\n');
};

/**
 * This function takes either a single CSS property or an array of CSS
 * properties & applies the provided value during the process, the value
 * is extracted from the provided themeSection if possible. If not, we assume
 * the user has provided a CSS value directly and fallback to that.
 */
const convertCssPropertiesToCssValues = (
  property: string | string[],
  value: DefaultThemeOrCSSProp<any, any> | Array<DefaultThemeOrCSSProp<any, any>>,
  themeSection: DefaultTheme[keyof DefaultTheme],
) => {
  if (Array.isArray(property) && Array.isArray(value)) {
    /**
     * If the value is an array & the cssProperty is an array, map the values to respective logical props.
     * This is normally the case for padding, margin, etc. where we extract each value and turn it into a single logical prop.
     *
     * @example cssProperty = ['padding-block-start', 'padding-inline-end', 'padding-block-end', 'padding-inline-start'] and breakpointValue = [3, 4]
     *
     * so this code would become
     * ```css
     * padding-block-start: 3;
     * padding-block-end: 3;
     * padding-inline-start: 4;
     * padding-inline-end: 4;
     * ```
     */
    const shorthandValues = fillCssValues(value);

    return property.reduce((acc, prop, index) => {
      acc[prop] = extractStyleFromTheme(themeSection, shorthandValues[index], shorthandValues[index]);

      return acc;
    }, {});
  } else if (Array.isArray(property) && !Array.isArray(value)) {
    /**
     * If the value is not an array & the cssProperty is an array, map the value to respective logical props (they'd all be the same).
     *
     * @example cssProperty = ['padding-block-start', 'padding-inline-end', 'padding-block-end', 'padding-inline-start'] and breakpointValue = 1
     *
     * so this code would become
     * ```css
     *  padding-inline-start: 1;
     *  padding-inline-end: 1;
     *  padding-block-start: 1;
     *  padding-block-end: 1;
     * ```
     */
    return property.reduce((acc, prop) => {
      acc[prop] = extractStyleFromTheme(themeSection, value, value);

      return acc;
    }, {});
  } else if (!Array.isArray(property) && !Array.isArray(value)) {
    return {
      [property]: extractStyleFromTheme(themeSection, value, value),
    };
  } else {
    console.warn(
      "You've passed an array of values to a property that does not support it. Please check the property and value you're passing.",
    );

    return {};
  }
};

export { handleResponsiveValues };
export type { ResponsiveProps, ResponsiveThemeProperty, ResponsiveProperty, Breakpoint };
