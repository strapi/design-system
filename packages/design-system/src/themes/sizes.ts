export interface Sizes {
  accordions: {
    S: string;
    M: string;
  };
  button: {
    S: string;
    M: string;
    L: string;
  };
}
export const sizes: Sizes = {
  accordions: {
    S: `4.8rem`,
    M: `8.8rem`,
  },
  button: {
    S: `3.2rem`,
    M: `3.6rem`,
    L: `4rem`,
  },
};
