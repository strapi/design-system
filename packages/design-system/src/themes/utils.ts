import { css, DefaultTheme, IStyledComponent } from 'styled-components';

export const getThemeSize = <TType extends keyof DefaultTheme['sizes']>(type: TType) => {
  return ({ theme, size }: { theme: DefaultTheme; size: keyof DefaultTheme['sizes'][TType] }) =>
    theme.sizes[type][size];
};

export const inputFocusStyle =
  (rootElement: IStyledComponent<'web'> | string = '&') =>
  ({ theme, $hasError = false }: { theme: DefaultTheme; $hasError?: boolean }) => css`
    outline: none;
    box-shadow: none;
    transition-property: border-color, box-shadow, fill;
    transition-duration: 0.2s;

    ${rootElement}:focus-within {
      border: 1px solid ${$hasError ? theme.colors.danger600 : theme.colors.primary600};
      box-shadow: ${$hasError ? theme.colors.danger600 : theme.colors.primary600} 0px 0px 0px 2px;
    }
  `;
