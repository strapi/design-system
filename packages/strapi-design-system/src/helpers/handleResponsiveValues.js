/* eslint-disable consistent-return */
const handleResponsiveValues = (property, value, theme) => {
  let transformedArray = value;

  if (!Array.isArray(value) && typeof value === 'object') {
    transformedArray = [value?.desktop, value?.tablet, value?.mobile];
  }

  if (transformedArray === undefined) return undefined;

  if (Array.isArray(transformedArray)) {
    const [desktopValue, tabletValue, mobileValue] = transformedArray;

    let spaces = `${property}: ${theme.spaces[desktopValue]};`;

    if (tabletValue !== undefined) {
      spaces += `${theme.mediaQueries.tablet}{
          ${property}: ${theme.spaces[tabletValue]};
        }`;
    }

    if (mobileValue !== undefined) {
      spaces += `${theme.mediaQueries.mobile}{
          ${property}: ${theme.spaces[mobileValue]};
        }`;
    }

    return spaces;
  }

  // Fallback to the passed transformedArray when necessary
  const realValue = theme.spaces[transformedArray] || transformedArray;

  return `${property}: ${realValue};`;
};

export default handleResponsiveValues;
