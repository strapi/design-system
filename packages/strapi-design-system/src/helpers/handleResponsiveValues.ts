import { DefaultTheme } from 'styled-components';

interface ResponsiveValueObject {
  desktop?: keyof DefaultTheme['spaces'];
  tablet?: keyof DefaultTheme['spaces'];
  mobile?: keyof DefaultTheme['spaces'];
}

type ResponsiveValueTuple = [
  desktop?: keyof DefaultTheme['spaces'],
  tablet?: keyof DefaultTheme['spaces'],
  mobile?: keyof DefaultTheme['spaces'],
];

export type ResponsiveValue = ResponsiveValueObject | keyof DefaultTheme['spaces'] | ResponsiveValueTuple;

/* eslint-disable consistent-return */
const handleResponsiveValues = (property: string, value: ResponsiveValue | undefined, theme: DefaultTheme) => {
  if (!value) {
    return undefined;
  }

  let transformedArray: ResponsiveValueTuple = Array.isArray(value) ? value : [];

  if (!Array.isArray(value) && typeof value === 'object') {
    transformedArray = [value?.desktop, value?.tablet, value?.mobile];
  }

  if (transformedArray.length > 0) {
    const spaces = transformedArray.reduce((acc, curr, index) => {
      if (curr) {
        switch (index) {
          case 0:
            return `${acc}${property}: ${theme.spaces[curr]};`;
          case 1:
            return `${acc}${theme.mediaQueries.tablet}{${property}: ${theme.spaces[curr]};}`;
          case 2:
            return `${acc}${theme.mediaQueries.mobile}{${property}: ${theme.spaces[curr]};}`;
          default:
            return acc;
        }
      }

      return acc;
    }, '');

    return spaces;
  }

  // Fallback to the passed transformedArray when necessary
  const realValue = theme.spaces[value as keyof DefaultTheme['spaces']] || value;

  return `${property}: ${realValue};`;
};

export default handleResponsiveValues;
