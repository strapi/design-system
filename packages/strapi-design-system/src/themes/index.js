export * from './lightTheme';
export * from './darkTheme';
export * from './extendTheme';
export * from './utils';
