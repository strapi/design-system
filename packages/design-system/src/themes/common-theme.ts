import { EASINGS, TIMINGS, TRANSITIONS } from '../styles/motion';

import { Sizes, sizes } from './sizes';

type Breakpoint = `@media(min-width: ${number}px)`;

export interface CommonTheme {
  sizes: Sizes;
  zIndices: {
    navigation: 100;
    overlay: 300;
    modal: 310;
    dialog: 320;
    popover: 500;
    notification: 700;
    tooltip: 1000;
  };
  spaces: ['0px', '4px', '8px', '12px', '16px', '20px', '24px', '32px', '40px', '48px', '56px', '64px'];
  breakpoints: {
    initial: Breakpoint;
    small: Breakpoint;
    medium: Breakpoint;
    large: Breakpoint;
  };
  borderRadius: '4px';
  fontSizes: [string, string, string, string, string, string];
  lineHeights: [1.14, 1.22, 1.25, 1.33, 1.43, 1.45, 1.5];
  fontWeights: {
    regular: 400;
    semiBold: 500;
    bold: 600;
  };
  motion: {
    easings: typeof EASINGS;
    timings: typeof TIMINGS;
  };
  transitions: typeof TRANSITIONS;
}

export const commonTheme: CommonTheme = {
  sizes,
  zIndices: {
    navigation: 100,
    overlay: 300,
    modal: 310,
    dialog: 320,
    popover: 500,
    notification: 700,
    tooltip: 1000,
  },
  spaces: ['0px', '4px', '8px', '12px', '16px', '20px', '24px', '32px', '40px', '48px', '56px', '64px'],
  borderRadius: '4px',
  breakpoints: {
    initial: '@media(min-width: 0px)',
    small: '@media(min-width: 520px)',
    medium: '@media(min-width: 768px)',
    large: '@media(min-width: 1280px)',
  },
  fontSizes: [`1.1rem`, `1.2rem`, `1.4rem`, '1.6rem', `1.8rem`, `3.2rem`],
  lineHeights: [1.14, 1.22, 1.25, 1.33, 1.43, 1.45, 1.5],
  fontWeights: {
    regular: 400,
    semiBold: 500,
    bold: 600,
  },
  transitions: TRANSITIONS,
  motion: {
    easings: EASINGS,
    timings: TIMINGS,
  },
};
