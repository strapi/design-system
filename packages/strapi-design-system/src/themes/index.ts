import { Colors, Shadows } from './colors';
import { CommonTheme } from './common-theme';

export * from './lightTheme';
export * from './darkTheme';
export * from './extendTheme';
export * from './utils';

export interface StrapiTheme extends CommonTheme {
  colors: Colors;
  shadows: Shadows;
}
