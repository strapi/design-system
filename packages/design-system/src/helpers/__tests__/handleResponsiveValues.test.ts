import { lightTheme } from '../../themes';
import { handleResponsiveValues } from '../handleResponsiveValues';

describe('handleResponsiveValues', () => {
  describe('spaces', () => {
    it('should handle simple string values', () => {
      expect(handleResponsiveValues({ padding: '1rem', margin: '4px' }, lightTheme)).toMatchInlineSnapshot(`
              "padding-block-start: 1rem;
              padding-inline-end: 1rem;
              padding-block-end: 1rem;
              padding-inline-start: 1rem;
              margin-block-start: 4px;
              margin-inline-end: 4px;
              margin-block-end: 4px;
              margin-inline-start: 4px;"
          `);
    });

    it('should handle simple space values from theme', () => {
      expect(handleResponsiveValues({ padding: 2, margin: 4 }, lightTheme)).toMatchInlineSnapshot(`
              "padding-block-start: 8px;
              padding-inline-end: 8px;
              padding-block-end: 8px;
              padding-inline-start: 8px;
              margin-block-start: 16px;
              margin-inline-end: 16px;
              margin-block-end: 16px;
              margin-inline-start: 16px;"
          `);
    });

    it('should handle responsive values with different breakpoints', () => {
      const result = handleResponsiveValues(
        {
          padding: {
            initial: 1,
            small: 2,
            medium: 3,
            large: 4,
          },
          margin: {
            initial: 2,
            small: 4,
            medium: 6,
            large: 8,
          },
        },
        lightTheme,
      );
      expect(result).toMatchInlineSnapshot(`
              "padding-block-start: 4px;
              padding-inline-end: 4px;
              padding-block-end: 4px;
              padding-inline-start: 4px;
              margin-block-start: 8px;
              margin-inline-end: 8px;
              margin-block-end: 8px;
              margin-inline-start: 8px;
              @media(min-width: 520px){ padding-block-start: 8px;
              padding-inline-end: 8px;
              padding-block-end: 8px;
              padding-inline-start: 8px;
              margin-block-start: 16px;
              margin-inline-end: 16px;
              margin-block-end: 16px;
              margin-inline-start: 16px; }
              @media(min-width: 768px){ padding-block-start: 12px;
              padding-inline-end: 12px;
              padding-block-end: 12px;
              padding-inline-start: 12px;
              margin-block-start: 24px;
              margin-inline-end: 24px;
              margin-block-end: 24px;
              margin-inline-start: 24px; }
              @media(min-width: 1280px){ padding-block-start: 16px;
              padding-inline-end: 16px;
              padding-block-end: 16px;
              padding-inline-start: 16px;
              margin-block-start: 40px;
              margin-inline-end: 40px;
              margin-block-end: 40px;
              margin-inline-start: 40px; }"
          `);
    });

    it('should handle responsive values with mixed prop type', () => {
      expect(
        handleResponsiveValues(
          {
            padding: {
              initial: 1,
              small: 2,
              medium: '1rem',
              large: 4,
            },
            // CSS specificity wont work here, so padding-right has to be after padding
            paddingRight: {
              initial: 2,
              small: 4,
              medium: 6,
              large: '8px',
            },
            margin: 2,
            marginTop: {
              initial: 1,
              small: '2rem',
            },
          },
          lightTheme,
        ),
      ).toMatchInlineSnapshot(`
              "padding-block-start: 4px;
              padding-inline-end: 8px;
              padding-block-end: 4px;
              padding-inline-start: 4px;
              margin-block-start: 4px;
              margin-inline-end: 8px;
              margin-block-end: 8px;
              margin-inline-start: 8px;
              @media(min-width: 520px){ padding-block-start: 8px;
              padding-inline-end: 16px;
              padding-block-end: 8px;
              padding-inline-start: 8px;
              margin-block-start: 2rem; }
              @media(min-width: 768px){ padding-block-start: 1rem;
              padding-inline-end: 24px;
              padding-block-end: 1rem;
              padding-inline-start: 1rem; }
              @media(min-width: 1280px){ padding-block-start: 16px;
              padding-inline-end: 8px;
              padding-block-end: 16px;
              padding-inline-start: 16px; }"
          `);
    });

    it('should handle responsive values with breakpoint array values', () => {
      expect(
        handleResponsiveValues(
          {
            padding: { initial: ['4rem'], medium: ['2rem', '4rem', '8rem'], large: 6 },
            margin: { initial: 2, medium: [2, '4px'], large: [6, '4px', 8] },
            marginTop: { initial: 4, medium: '1rem', large: '22px' },
          },
          lightTheme,
        ),
      ).toMatchInlineSnapshot(`
              "padding-block-start: 4rem;
              padding-inline-end: 4rem;
              padding-block-end: 4rem;
              padding-inline-start: 4rem;
              margin-block-start: 16px;
              margin-inline-end: 8px;
              margin-block-end: 8px;
              margin-inline-start: 8px;
              @media(min-width: 768px){ padding-block-start: 2rem;
              padding-inline-end: 4rem;
              padding-block-end: 8rem;
              padding-inline-start: 4rem;
              margin-block-start: 1rem;
              margin-inline-end: 4px;
              margin-block-end: 8px;
              margin-inline-start: 4px; }
              @media(min-width: 1280px){ padding-block-start: 24px;
              padding-inline-end: 24px;
              padding-block-end: 24px;
              padding-inline-start: 24px;
              margin-block-start: 22px;
              margin-inline-end: 4px;
              margin-block-end: 40px;
              margin-inline-start: 4px; }"
          `);
    });

    it('should handle simple value array and map to logical props wherever required', () => {
      expect(
        handleResponsiveValues(
          {
            padding: ['2rem', '4rem', '8rem'],
            margin: [2, '4px'],
            paddingTop: { initial: '4rem', medium: 4, large: 6 },
            marginTop: 1,
          },
          lightTheme,
        ),
      ).toMatchInlineSnapshot(`
              "padding-block-start: 4rem;
              padding-inline-end: 4rem;
              padding-block-end: 8rem;
              padding-inline-start: 4rem;
              margin-block-start: 4px;
              margin-inline-end: 4px;
              margin-block-end: 8px;
              margin-inline-start: 4px;
              @media(min-width: 768px){ padding-block-start: 16px; }
              @media(min-width: 1280px){ padding-block-start: 24px; }"
          `);
    });
  });

  describe('colors', () => {
    it('should handle simple color values', () => {
      expect(handleResponsiveValues({ background: 'neutral600', borderColor: 'warning600' }, lightTheme))
        .toMatchInlineSnapshot(`
        "background: #666687;
        border-color: #d9822f;"
      `);
    });

    it('should handle responsive values with different breakpoints', () => {
      const result = handleResponsiveValues(
        {
          background: {
            initial: 'primary600',
            small: 'secondary600',
            medium: 'neutral600',
            large: 'neutral800',
          },
          color: {
            initial: 'danger600',
            small: 'pink',
            // @ts-expect-error at "large: 3", Type 'number' is not assignable to type 'string'
            large: 3,
          },
        },
        lightTheme,
      );
      expect(result).toMatchInlineSnapshot(`
        "background: #4945ff;
        color: #d02b20;
        @media(min-width: 520px){ background: #0c75af;
        color: pink; }
        @media(min-width: 768px){ background: #666687; }
        @media(min-width: 1280px){ background: #32324d;
        color: 3; }"
      `);
    });
  });

  describe('gap', () => {
    it('should handle simple gap values', () => {
      expect(handleResponsiveValues({ gap: 4 }, lightTheme)).toMatchInlineSnapshot(`"gap: 16px;"`);
    });

    it('should handle responsive values with different breakpoints', () => {
      expect(
        handleResponsiveValues(
          {
            gap: {
              initial: 1,
              small: 2,
              medium: 3,
              large: '24px',
            },
          },
          lightTheme,
        ),
      ).toMatchInlineSnapshot(`
        "gap: 4px;
        @media(min-width: 520px){ gap: 8px; }
        @media(min-width: 768px){ gap: 12px; }
        @media(min-width: 1280px){ gap: 24px; }"
      `);
    });
  });

  describe('font sizes', () => {
    it('should handle simple fontSize values', () => {
      expect(handleResponsiveValues({ fontSize: 2 }, lightTheme)).toMatchInlineSnapshot(`"font-size: 1.4rem;"`);
    });

    it('should handle responsive values with different breakpoints', () => {
      expect(
        handleResponsiveValues(
          {
            fontSize: {
              initial: 2,
              small: 3,
              medium: '14px',
              large: 5,
            },
          },
          lightTheme,
        ),
      ).toMatchInlineSnapshot(`
        "font-size: 1.4rem;
        @media(min-width: 520px){ font-size: 1.6rem; }
        @media(min-width: 768px){ font-size: 14px; }
        @media(min-width: 1280px){ font-size: 3.2rem; }"
      `);
    });
  });

  describe('font weights', () => {
    it('should handle simple fontWeight values', () => {
      expect(handleResponsiveValues({ fontWeight: 'bold' }, lightTheme)).toMatchInlineSnapshot(`"font-weight: 600;"`);
    });

    it('should handle responsive values with different breakpoints', () => {
      expect(
        handleResponsiveValues(
          {
            fontWeight: {
              initial: 'regular',
              small: 'semiBold',
              medium: 'bold',
              large: 850,
            },
          },
          lightTheme,
        ),
      ).toMatchInlineSnapshot(`
        "font-weight: 400;
        @media(min-width: 520px){ font-weight: 500; }
        @media(min-width: 768px){ font-weight: 600; }
        @media(min-width: 1280px){ font-weight: 850; }"
      `);
    });
  });

  describe('line heights', () => {
    it('should handle simple lineHeight values', () => {
      expect(handleResponsiveValues({ lineHeight: 1.5 }, lightTheme)).toMatchInlineSnapshot(`"line-height: 1.5;"`);
    });

    it('should handle responsive values with different breakpoints', () => {
      expect(
        handleResponsiveValues(
          {
            lineHeight: {
              initial: 2,
              small: 1.22,
              medium: 3,
              large: 'invalidValue',
            },
          },
          lightTheme,
        ),
      ).toMatchInlineSnapshot(`
        "line-height: 1.25;
        @media(min-width: 520px){ line-height: 1.22; }
        @media(min-width: 768px){ line-height: 1.33; }
        @media(min-width: 1280px){ line-height: invalidValue; }"
      `);
    });
  });

  describe('zIndices', () => {
    it('should handle simple zIndex values', () => {
      expect(handleResponsiveValues({ zIndex: 'modal' }, lightTheme)).toMatchInlineSnapshot(`"z-index: 310;"`);
    });

    it('should handle responsive values with different breakpoints', () => {
      expect(
        handleResponsiveValues(
          {
            zIndex: {
              initial: 'navigation',
              small: 9,
              medium: 'tooltip',
              large: 'invalidValue',
            },
          },
          lightTheme,
        ),
      ).toMatchInlineSnapshot(`
        "z-index: 100;
        @media(min-width: 520px){ z-index: 9; }
        @media(min-width: 768px){ z-index: 1000; }
        @media(min-width: 1280px){ z-index: invalidValue; }"
      `);
    });
  });

  describe('shadows', () => {
    it('should handle simple shadow values', () => {
      expect(handleResponsiveValues({ boxShadow: 'filterShadow' }, lightTheme)).toMatchInlineSnapshot(
        `"box-shadow: 0px 1px 4px rgba(33, 33, 52, 0.1);"`,
      );
    });

    it('should handle responsive values with different breakpoints', () => {
      expect(
        handleResponsiveValues(
          {
            boxShadow: {
              initial: 'filterShadow',
              small: 'focusShadow',
              // @ts-expect-error at "medium: 3", Type 'number' is not assignable to type 'string'
              medium: 3,
              large: '0px 1px 4px rgba(33, 33, 52, 0.1)',
            },
          },
          lightTheme,
        ),
      ).toMatchInlineSnapshot(`
        "box-shadow: 0px 1px 4px rgba(33, 33, 52, 0.1);
        @media(min-width: 520px){ box-shadow: 0px 0px 6px rgba(76, 191, 255, 0.75); }
        @media(min-width: 768px){ box-shadow: 3; }
        @media(min-width: 1280px){ box-shadow: 0px 1px 4px rgba(33, 33, 52, 0.1); }"
      `);
    });
  });

  describe('text decoration', () => {
    it('should handle simple text decoration values', () => {
      const values = { textDecoration: 'underline' };
      expect(handleResponsiveValues(values, lightTheme)).toMatchInlineSnapshot(`"text-decoration: underline;"`);
    });

    it('should handle responsive values with different breakpoints', () => {
      const values = {
        textDecoration: {
          initial: 'none',
          small: 'underline',
          medium: 'overline',
        },
      };
      expect(handleResponsiveValues(values, lightTheme)).toMatchInlineSnapshot(`
        "text-decoration: none;
        @media(min-width: 520px){ text-decoration: underline; }
        @media(min-width: 768px){ text-decoration: overline; }"
      `);
    });
  });
});
