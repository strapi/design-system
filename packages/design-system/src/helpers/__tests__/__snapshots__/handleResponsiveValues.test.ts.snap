// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleResponsiveValues should handle responsive values with breakpoint array values 1`] = `
"padding-block-start: 4rem;
padding-inline-end: 4rem;
padding-block-end: 4rem;
padding-inline-start: 4rem;
margin-block-start: 16px;
margin-inline-end: 8px;
margin-block-end: 8px;
margin-inline-start: 8px;
@media(min-width: 768px){ padding-block-start: 2rem;
padding-inline-end: 4rem;
padding-block-end: 8rem;
padding-inline-start: 4rem;
margin-block-start: 1rem;
margin-inline-end: 4px;
margin-block-end: 8px;
margin-inline-start: 4px; }
@media(min-width: 1280px){ padding-block-start: 24px;
padding-inline-end: 24px;
padding-block-end: 24px;
padding-inline-start: 24px;
margin-block-start: 22px;
margin-inline-end: 4px;
margin-block-end: 40px;
margin-inline-start: 4px; }"
`;

exports[`handleResponsiveValues should handle responsive values with different breakpoints 1`] = `
"padding-block-start: 4px;
padding-inline-end: 4px;
padding-block-end: 4px;
padding-inline-start: 4px;
margin-block-start: 8px;
margin-inline-end: 8px;
margin-block-end: 8px;
margin-inline-start: 8px;
@media(min-width: 520px){ padding-block-start: 8px;
padding-inline-end: 8px;
padding-block-end: 8px;
padding-inline-start: 8px;
margin-block-start: 16px;
margin-inline-end: 16px;
margin-block-end: 16px;
margin-inline-start: 16px; }
@media(min-width: 768px){ padding-block-start: 12px;
padding-inline-end: 12px;
padding-block-end: 12px;
padding-inline-start: 12px;
margin-block-start: 24px;
margin-inline-end: 24px;
margin-block-end: 24px;
margin-inline-start: 24px; }
@media(min-width: 1280px){ padding-block-start: 16px;
padding-inline-end: 16px;
padding-block-end: 16px;
padding-inline-start: 16px;
margin-block-start: 40px;
margin-inline-end: 40px;
margin-block-end: 40px;
margin-inline-start: 40px; }"
`;

exports[`handleResponsiveValues should handle responsive values with mixed prop type 1`] = `
"padding-block-start: 4px;
padding-inline-end: 8px;
padding-block-end: 4px;
padding-inline-start: 4px;
margin-block-start: 4px;
margin-inline-end: 8px;
margin-block-end: 8px;
margin-inline-start: 8px;
@media(min-width: 520px){ padding-block-start: 8px;
padding-inline-end: 16px;
padding-block-end: 8px;
padding-inline-start: 8px;
margin-block-start: 2rem; }
@media(min-width: 768px){ padding-block-start: 1rem;
padding-inline-end: 24px;
padding-block-end: 1rem;
padding-inline-start: 1rem; }
@media(min-width: 1280px){ padding-block-start: 16px;
padding-inline-end: 8px;
padding-block-end: 16px;
padding-inline-start: 16px; }"
`;

exports[`handleResponsiveValues should handle simple space values from theme 1`] = `
"padding-block-start: 8px;
padding-inline-end: 8px;
padding-block-end: 8px;
padding-inline-start: 8px;
margin-block-start: 16px;
margin-inline-end: 16px;
margin-block-end: 16px;
margin-inline-start: 16px;"
`;

exports[`handleResponsiveValues should handle simple string values 1`] = `
"padding-block-start: 1rem;
padding-inline-end: 1rem;
padding-block-end: 1rem;
padding-inline-start: 1rem;
margin-block-start: 4px;
margin-inline-end: 4px;
margin-block-end: 4px;
margin-inline-start: 4px;"
`;

exports[`handleResponsiveValues should handle simple value array and map to logical props wherever required 1`] = `
"padding-block-start: 4rem;
padding-inline-end: 4rem;
padding-block-end: 8rem;
padding-inline-start: 4rem;
margin-block-start: 4px;
margin-inline-end: 4px;
margin-block-end: 8px;
margin-inline-start: 4px;
@media(min-width: 768px){ padding-block-start: 16px; }
@media(min-width: 1280px){ padding-block-start: 24px; }"
`;
